### PR TITLE
fix(statistics): #MA-1038 save for each student instead of each strcture

### PR DESCRIPTION
**Pull request by BastienBenard**
## Describe your changes
When the worker calculated the statistics, we saved the student data for each structure.
If we had 1000 students who each had 1000 events, we would end up with a large amount of data to save in one request.
This could cause timeouts in the mongo.
To solve this problem, we make a backup request per student instead of one per structure.

The tests will be done in issue #248

## Issue ticket number and link
[MA-1038](https://jira.support-ent.fr/browse/MA-1038)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: statistics-presences/src/main/java/fr/openent/statistics_presences/indicator/IndicatorWorker.java [hunk 1]
```diff
@@ -100,6 +100,23 @@ protected void save(String id, JsonArray students, List<JsonObject> values, Hand
         this.save(id, students, values, null, null, handler);
     }
 
+    protected Future<List<JsonObject>> saveStudent(String structureId, String studentId, List<JsonObject> values, String startDate,
```
Comment (sjollois) [inline statistics-presences/src/main/java/fr/openent/statistics_presences/indicator/IndicatorWorker.java]: pk la fonction saveStudent fait un deleteOldValuesForStudent ?? il y a une incohérence au niveau des noms

**Review by sjollois — CHANGES_REQUESTED** [review]

**Review by sjollois — APPROVED** [review]

File: statistics-presences/src/main/java/fr/openent/statistics_presences/indicator/IndicatorWorker.java [hunk 2]
```diff
@@ -161,6 +178,36 @@ private Future<List<JsonObject>> deleteOldValues(String id, JsonArray students,
...
 
+    private Future<List<JsonObject>> deleteOldValuesForStudent(String id, String studentId, List<JsonObject> values, String startDate, String endDate) {
+        Future<List<JsonObject>> future = Future.future();
+        JsonObject selector = new JsonObject()
+                .put(Field.INDICATOR, this.indicatorName())
+                .put(Field.STRUCTURE, id)
+                .put(Field.USER, studentId);
+        if (startDate != null && endDate != null) {
+            JsonObject $gte = new JsonObject()
+                    .put(Field.$GTE, startDate);
```
Comment (flmartineau) [inline statistics-presences/src/main/java/fr/openent/statistics_presences/indicator/IndicatorWorker.java:189]: Pas obligé de le faire maintenant, mais à terme il faudrait un fichier constants pour les clés mongo

File: statistics-presences/src/main/java/fr/openent/statistics_presences/indicator/IndicatorWorker.java [hunk 3]
```diff
@@ -345,42 +390,29 @@ private Future<List<JsonObject>> processStudent(String structureId, String stude
...
+            return new ArrayList<>();
         }
+
+        return statsByEventTypes.entrySet().stream()
+                .flatMap(statsByEventType -> statsByEventType.getValue().result().stream()
+                        .map(stat -> {
+                            stat.setUser(studentId)
+                                    .setIndicator(indicatorName())
+                                    .setName(student.getString("name"))
+                                    .setClassName(String.join(",", student.getJsonArray("className").getList()))
```
Comment (flmartineau) [inline statistics-presences/src/main/java/fr/openent/statistics_presences/indicator/IndicatorWorker.java]: Magic strings

**Review by flmartineau — APPROVED** [review]

File: statistics-presences/src/main/java/fr/openent/statistics_presences/indicator/IndicatorWorker.java (hunk 2 above)
Comment (BastienBenard) [reply statistics-presences/src/main/java/fr/openent/statistics_presences/indicator/IndicatorWorker.java:189]: j'ai crée cette issu #250

**Pull request merged**